plugins/usbdmx/AsyncUsbReceiver.cpp

```diff
@@ -0,0 +1,182 @@
+/*
+ * This program is free software; you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License as published by
+ * the Free Software Foundation; either version 2 of the License, or
+ * (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU Library General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program; if not, write to the Free Software
+ * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301 USA.
+ *
+ * AsyncUsbReceiver.cpp
+ * An Asynchronous DMX USB receiver.
+ * Copyright (C) 2015 Stefan Krupop
+ */
+
+#include "plugins/usbdmx/AsyncUsbReceiver.h"
+
+#include "ola/Logging.h"
+
+namespace ola {
+namespace plugin {
+namespace usbdmx {
+
+namespace {
+
+/*
+ * Called by libusb when the transfer completes.
+ */
+#ifdef _WIN32
+__attribute__((__stdcall__))
+#endif
+void AsyncCallback(struct libusb_transfer *transfer) {
+  AsyncUsbReceiver *widget = reinterpret_cast<AsyncUsbReceiver*>(
+      transfer->user_data);
+  widget->TransferComplete(transfer);
+}
+}  // namespace
+
+AsyncUsbReceiver::AsyncUsbReceiver(ola::usb::LibUsbAdaptor *adaptor,
+                                   libusb_device *usb_device,
+                                   PluginAdaptor *plugin_adaptor)
+    : m_adaptor(adaptor),
+      m_usb_device(usb_device),
+      m_usb_handle(NULL),
+      m_plugin_adaptor(plugin_adaptor),
+      m_inited_with_handle(false),
+      m_suppress_continuation(false),
+      m_transfer_state(IDLE) {
+  m_transfer = m_adaptor->AllocTransfer(0);
+  m_adaptor->RefDevice(usb_device);
+}
+
+AsyncUsbReceiver::~AsyncUsbReceiver() {
+  CancelTransfer();
+  if (!m_inited_with_handle) {
+    m_adaptor->Close(m_usb_handle);
+  }
+  m_adaptor->UnrefDevice(m_usb_device);
+  m_adaptor->FreeTransfer(m_transfer);
+}
+
+bool AsyncUsbReceiver::Init() {
+  m_usb_handle = SetupHandle();
+  m_inited_with_handle = false;
+  return m_usb_handle ? true : false;
+}
+
+bool AsyncUsbReceiver::Init(libusb_device_handle* handle) {
+  m_usb_handle = handle;
+  m_inited_with_handle = true;
+  return true;
+}
+
+bool AsyncUsbReceiver::Start() {
```

Comment:
Ok, I created a base class "AsyncUsbTransceiverBase" and moved the common parts there.